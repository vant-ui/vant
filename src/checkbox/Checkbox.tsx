import { computed, watch, defineComponent } from 'vue';

// Utils
import { createNamespace, pick } from '../utils';
import {
  CHECKBOX_GROUP_KEY,
  CheckboxGroupProvide,
} from '../checkbox-group/CheckboxGroup';

// Composables
import { useParent } from '@vant/use';
import { useExpose } from '../composables/use-expose';
import { useLinkField } from '../composables/use-link-field';

// Components
import Checker, { checkerProps } from './Checker';

const [name, bem] = createNamespace('checkbox');

export default defineComponent({
  name,

  props: {
    ...checkerProps,
    bindGroup: {
      type: Boolean,
      default: true,
    },
  },

  emits: ['change', 'update:modelValue'],

  setup(props, { emit, slots }) {
    const { parent } = useParent<CheckboxGroupProvide>(CHECKBOX_GROUP_KEY);

    const setParentValue = (checked: boolean) => {
      const { name } = props;
      const { max, modelValue } = parent!.props;
      const value = modelValue.slice();

      if (checked) {
        const overlimit = max && value.length >= max;

        if (!overlimit && !value.includes(name)) {
          value.push(name);

          if (props.bindGroup) {
            parent!.updateValue(value);
          }
        }
      } else {
        const index = value.indexOf(name);

        if (index !== -1) {
          value.splice(index, 1);

          if (props.bindGroup) {
            parent!.updateValue(value);
          }
        }
      }
    };

    const checked = computed(() => {
      if (parent && props.bindGroup) {
        return parent.props.modelValue.indexOf(props.name) !== -1;
      }
      return !!props.modelValue;
    });

    const toggle = (newValue = !checked.value) => {
      if (parent && props.bindGroup) {
        setParentValue(newValue);
      } else {
        emit('update:modelValue', newValue);
      }
    };

    watch(
      () => props.modelValue,
      (value) => emit('change', value)
    );

    useExpose({ toggle, props, checked });
    useLinkField(() => props.modelValue);

    return () => (
      <Checker
        v-slots={pick(slots, ['default', 'icon'])}
        bem={bem}
        role="checkbox"
        parent={parent}
        checked={checked.value}
        onToggle={toggle}
        {...props}
      />
    );
  },
});
