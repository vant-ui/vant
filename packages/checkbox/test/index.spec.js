import Checkbox from '../';
import demoTest from '../../../test/demo-test';

demoTest(Checkbox);
