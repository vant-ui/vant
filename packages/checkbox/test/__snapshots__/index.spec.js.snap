// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders checkbox correctly 1`] = `
<section class="van-doc-demo-section demo-checkbox" style="background:undefined;">
  <section class="van-doc-demo-block">
    <h2 class="van-doc-demo-block__title">基础用法</h2>
    <div class="van-checkbox">
      <i class="van-icon van-icon-success van-checkbox__icon van-checkbox--round van-checkbox--checked">
        <!---->
      </i> <span class="van-checkbox__label">复选框</span></div>
  </section>
  <section class="van-doc-demo-block">
    <h2 class="van-doc-demo-block__title">禁用状态</h2>
    <div class="van-checkbox">
      <i class="van-icon van-icon-success van-checkbox__icon van-checkbox--round van-checkbox--disabled">
        <!---->
      </i> <span class="van-checkbox__label">复选框</span></div>
    <div class="van-checkbox">
      <i class="van-icon van-icon-success van-checkbox__icon van-checkbox--round van-checkbox--disabled van-checkbox--checked">
        <!---->
      </i> <span class="van-checkbox__label">复选框</span></div>
  </section>
  <section class="van-doc-demo-block">
    <h2 class="van-doc-demo-block__title">禁用内容部分点击事件</h2>
    <div class="van-checkbox">
      <i class="van-icon van-icon-success van-checkbox__icon van-checkbox--round van-checkbox--checked">
        <!---->
      </i> <span class="van-checkbox__label">复选框</span></div>
  </section>
  <section class="van-doc-demo-block">
    <h2 class="van-doc-demo-block__title">Checkbox 组</h2>
    <div class="van-checkbox-group">
      <div class="van-checkbox">
        <i class="van-icon van-icon-success van-checkbox__icon van-checkbox--round van-checkbox--checked">
          <!---->
        </i> <span class="van-checkbox__label">
        复选框 a
      </span></div>
      <div class="van-checkbox">
        <i class="van-icon van-icon-success van-checkbox__icon van-checkbox--round van-checkbox--checked">
          <!---->
        </i> <span class="van-checkbox__label">
        复选框 b
      </span></div>
      <div class="van-checkbox">
        <i class="van-icon van-icon-success van-checkbox__icon van-checkbox--round">
          <!---->
        </i> <span class="van-checkbox__label">
        复选框 c
      </span></div>
    </div>
  </section>
  <section class="van-doc-demo-block">
    <h2 class="van-doc-demo-block__title">设置最大可选数</h2>
    <div class="van-checkbox-group">
      <div class="van-checkbox">
        <i class="van-icon van-icon-success van-checkbox__icon van-checkbox--round">
          <!---->
        </i> <span class="van-checkbox__label">
        复选框 a
      </span></div>
      <div class="van-checkbox">
        <i class="van-icon van-icon-success van-checkbox__icon van-checkbox--round">
          <!---->
        </i> <span class="van-checkbox__label">
        复选框 b
      </span></div>
      <div class="van-checkbox">
        <i class="van-icon van-icon-success van-checkbox__icon van-checkbox--round">
          <!---->
        </i> <span class="van-checkbox__label">
        复选框 c
      </span></div>
    </div>
  </section>
  <section class="van-doc-demo-block">
    <h2 class="van-doc-demo-block__title">与 Cell 组件一起使用</h2>
    <div class="van-checkbox-group">
      <div class="van-cell-group van-hairline--top-bottom">
        <div class="van-cell van-hairline">
          <!---->
          <div class="van-cell__title"><span>复选框a</span>
            <!---->
          </div>
          <div class="van-cell__value">
            <div class="van-checkbox">
              <i class="van-icon van-icon-success van-checkbox__icon van-checkbox--round">
                <!---->
              </i>
              <!---->
            </div>
          </div>
          <!---->
        </div>
        <div class="van-cell van-hairline">
          <!---->
          <div class="van-cell__title"><span>复选框b</span>
            <!---->
          </div>
          <div class="van-cell__value">
            <div class="van-checkbox">
              <i class="van-icon van-icon-success van-checkbox__icon van-checkbox--round">
                <!---->
              </i>
              <!---->
            </div>
          </div>
          <!---->
        </div>
        <div class="van-cell van-hairline">
          <!---->
          <div class="van-cell__title"><span>复选框c</span>
            <!---->
          </div>
          <div class="van-cell__value">
            <div class="van-checkbox">
              <i class="van-icon van-icon-success van-checkbox__icon van-checkbox--round">
                <!---->
              </i>
              <!---->
            </div>
          </div>
          <!---->
        </div>
      </div>
    </div>
  </section>
</section>
`;
